p2p/src/net/mock/transport/adapted.rs

```diff
@@ -0,0 +1,138 @@
+            stream_adapter,
+            base_transport,
+        }
+    }
+
+    async fn bind(&self, address: Self::Address) -> Result<Self::Listener> {
+        AdaptedListener::start(
+            &self.base_transport,
+            Arc::clone(&self.stream_adapter),
+            address,
+        )
+        .await
+    }
+
+    async fn connect(&self, address: Self::Address) -> Result<Self::Stream> {
+        let base = self.base_transport.connect(address).await?;
+        let stream = self.stream_adapter.handshake(base, Role::Outbound).await?;
+        Ok(stream)
+    }
+}
+
+pub struct AdaptedListener<S: StreamAdapter<T::Stream>, T: MockTransport> {
+    receiver: UnboundedReceiver<(S::Stream, T::Address)>,
+    local_address: T::Address,
+    join_handle: tokio::task::JoinHandle<()>,
+}
+
+impl<S: StreamAdapter<T::Stream>, T: MockTransport> AdaptedListener<S, T> {
+    async fn start(transport: &T, stream_adapter: Arc<S>, address: T::Address) -> Result<Self> {
+        let mut listener = transport.bind(address).await?;
+        let local_address = listener.local_address()?;
+        let (sender, receiver) = tokio::sync::mpsc::unbounded_channel();
+
+        // Process new connections in background because MockListener::accept must be cancel safe.
+        let join_handle = tokio::spawn(async move {
+            loop {
+                let (base, addr) = match listener.accept().await {
+                    Ok(stream) => stream,
+                    Err(err) => {
+                        logging::log::error!("TCP accept failed unexpectedly: {}", err);
+                        return;
+                    }
+                };
+                let sender = sender.clone();
+                let stream_adapter_copy = Arc::clone(&stream_adapter);
+                tokio::spawn(async move {
+                    let res = timeout(
+                        HANDSHAKE_TIMEOUT,
+                        stream_adapter_copy.handshake(base, Role::Inbound),
+                    )
+                    .await;
+                    let socket = match res {
+                        Ok(Ok(socket)) => socket,
+                        Ok(Err(err)) => {
+                            logging::log::warn!("encryption handshake failed: {}", err);
+                            return;
+                        }
+                        Err(err) => {
+                            logging::log::warn!("encryption handshake timed out: {}", err);
+                            return;
+                        }
+                    };
+                    // It's not an error if the channel is already closed
+                    _ = sender.send((socket, addr));
+                });
+            }
+        });
+
+        Ok(Self {
+            receiver,
+            local_address,
+            join_handle,
+        })
+    }
+}
+
+impl<S: StreamAdapter<T::Stream>, T: MockTransport> Drop for AdaptedListener<S, T> {
+    fn drop(&mut self) {
+        self.join_handle.abort();
```

Comment:
Is aborting join handles the correct way to do this? Usually you want to wait for it to finish or something. I'm not sure.